domain/port/state/state.go

```diff
@@ -211,86 +190,164 @@ AND endpoint IN ($endpoints[:])
-	for _, protocol := range protocols {
-		protocolMap[protocol.Name] = protocol.ID
+func verifyNoPortRangeConflicts(currentOpenedPorts []endpointPortRangeUUID, inputPortRanges []network.PortRange) error {
+	for _, inputPortRange := range inputPortRanges {
+		for _, openedPortRange := range currentOpenedPorts {
+			if inputPortRange == openedPortRange.decode() {
+				// We allow port ranges to conflict only if they are equal
+				continue
+			}
+			if inputPortRange.ConflictsWith(openedPortRange.decode()) {
+				return errors.Annotatef(ErrPortRangeConflict,
+					"port range %q conflicts with existing port range %q",
+					inputPortRange, openedPortRange.decode())
+			}
+		}
 	}
-
-	return protocolMap, nil
+	return nil
 }
 
-// getOpenedPortsForEndpoints returns the opened ports for the given endpoints
-// of a given unit.
-func (st *State) getOpenedPortsForEndpoints(
-	ctx context.Context, tx *sqlair.TX, unitUUID unitUUID, endpoints endpoints,
-) (network.GroupedPortRanges, error) {
-	getOpenedPortsOnEndpoints, err := st.Prepare(`
-SELECT &endpointPortRange.*
-FROM port_range
-JOIN protocol ON port_range.protocol_id = protocol.id
-JOIN unit_endpoint ON port_range.unit_endpoint_uuid = unit_endpoint.uuid
-WHERE unit_endpoint.unit_uuid = $unitUUID.unit_uuid
-AND unit_endpoint.endpoint IN ($endpoints[:])
-`, endpointPortRange{}, unitUUID, endpoints)
+// openPorts inserts the given port ranges into the database, unless they're already open.
+func (st *State) openPorts(
+	ctx context.Context, tx *sqlair.TX,
+	openPorts network.GroupedPortRanges, currentOpenedPorts []endpointPortRangeUUID, endpoints []endpoint,
+) error {
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
 	if err != nil {
-		return nil, errors.Annotate(err, "preparing get opened ports on endpoints statement")
+		return errors.Annotate(err, "preparing insert port range statement")
 	}
 
-	var currentOpenedEndpointPorts []endpointPortRange
-	err = tx.Query(ctx, getOpenedPortsOnEndpoints, unitUUID, endpoints).GetAll(&currentOpenedEndpointPorts)
-	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
-		return nil, errors.Trace(err)
+	protocolMap, err := st.getProtocolMap(ctx, tx)
+	if err != nil {
+		return errors.Annotate(err, "getting protocol map")
 	}
 
-	endpointToPorts := make(network.GroupedPortRanges)
-	for _, r := range currentOpenedEndpointPorts {
-		endpointToPorts[r.Endpoint] = endpointToPorts[r.Endpoint].Add(r.decode())
+	// Construct the new port ranges to open
+	var openPortRanges []unitPortRange
+	for _, ep := range endpoints {
+		ports, ok := openPorts[ep.Endpoint]
+		if !ok {
+			continue
+		}
+		for _, portRange := range ports {
+			// skip port range if it's already open on this endpoint
+			found := false
+			for _, openedPortRange := range currentOpenedPorts {
+				if portRange == openedPortRange.decode() && ep.Endpoint == openedPortRange.Endpoint {
+					found = true
+					break
+				}
+			}
```

Comment:
To prevent for loop in a for loop in a for loop (accidental exponential), pre-calculate the currentOpenPorts once, then it's just an if check against a map.

Reply:
Done. I have pre-calculated a map which I lookup to find matched instead of embedding a for loop